accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+- Providing a query endpoint for retrieving attestations
+
+We're planning on providing detailed documentation on the required changes.
+
+### Non-falsifiable provenance using a trusted builder
+To ensure the provenance information hasn't been tampered with during the build, e.g. by a malicious dependency, we need to have a way to run a build and publish in a trusted build environment that can't be directly accessed or modified by the source repository where it's being executed.
+
+Non-falsifiable provenance information is a [requirement to reach SLSA level 3](https://slsa.dev/spec/v0.1/requirements#non-falsifiable) (framework for securing the supply chain).
+
+Using a trusted builder to publish an npm release will ensure a higher level of build integrity. Using it will be optional, but highly recommended. Verification of build provenance will include checks to determine if a trusted builder was used.
+
+At a high level, the trusted builder will checkout the git repo, install dependencies and execute `npm run build` followed by `npm publish`. Maintainers can only control what is run during `npm run build` not add or remove steps from the trusted builder. Provenance generation will be run in a separate workflow step that's isolated from the build and install.
+
+Using the trusted builder does not guarantee that your project has reached SLSA level 3. This could be achieved on a case-by-case basis by auditing the build scripts, making sure no untrusted scripts are fetched from outside the source repo and executed.
+
+#### Trusted builders on supported CI/CD systems
+We aim to offer a "trusted builder" (immutable build system) for npm on all CI/CD systems that support this pattern. Today, Circle ([Orbs](https://circleci.com/orbs/)) and GitHub Actions ([reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows)) can be used to implement a trusted builder. GitLab's supply chain vision includes adding [native support for SLSA](https://about.gitlab.com/direction/supply-chain/#frameworks).
+
+The [slsa-github-generator](https://github.com/slsa-framework/slsa-github-generator) is an example of a trusted builder using GitHub Actions [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows).
+
+### Provenance attestations
+The process of creating a provenance attestation is dependent on the current CI/CD system being used as there's currently no standardized way of retrieving this information. Different methods and variables will need to be used to gather information about the current branch, commit version, job id, actor etc.
+
+Existing tools like [witness](https://github.com/testifysec/witness) and Sigstore's [cosign](https://github.com/sigstore/cosign) CLI implement direct support for various CI/CD systems, so the [sigstore-js](https://github.com/sigstore/sigstore-js) client used in the npm CLI should follow that pattern and be able to figure out in which environment it’s running in.
```

Comment:
Should this be a hard blocker for landing sigstore-js?

Reply:
Sorry, to be more clear: should the direct support for various CI/CD systems that matches what witness and cosign do be a hard requirement prior to landing it in the npm CLI?

Reply:
The bar for this is low atm, only GitHub Actions is supported by Fulcio at the moment so the CLI should match this and keep up with any new CI/CD providers that are supported by Fulcio as they get added to https://github.com/sigstore/fulcio/tree/main/federation

Reply:
Clarified this 👍